src/main/java/org/apache/commons/imaging/formats/png/chunks/ChunkLoader.java

```diff
@@ -0,0 +1,322 @@
+    }
+    
+    /**
+     * ChunkSelector object filter out chunks that are not to be
+     * loaded by {@code ChunkLoader.run()} to reduce overall
+     * overhead.
+     * 
+     * <h2>NOTICE</h2>
+     * 
+     * (Replaces PngImageParser.keepChunk}. ChunkSelector also
+     * supports two mode - whiteList and blackList. In whiteList,
+     * the chunk types are the ones to be loaded, and in blackList
+     * all chunks are loaded except the given ones.
+     * 
+     * <h2>USAGE</h2>
+     * There are two ChunkSelector implementations - ChunkListSelector
+     * and ChunkUniselector.
+     */
+    public static abstract class ChunkSelector {
+        
+        protected final boolean whiteList;
+
+        /**
+         * Constructs a new ChunkSelector.
+         * 
+         * @param whiteList - whether chunk types given are to be
+         *      accepted or refused
+         */
+        protected ChunkSelector(boolean whiteList) {
+            this.whiteList = whiteList;
+        }
+        
+        public boolean isWhiteList() {
+            return whiteList;
+        }
+        
+        public boolean isBlackList() {
+            return !whiteList;
+        }
+        
+        /**
+         * Returns whether the given chunk is to be selected for
+         * building.
+         * 
+         * @param chunkType
+         * @return 
+         */
+        public abstract boolean whetherToKeep(ChunkType chunkType);
+        
+        /**
+         * Filters all the chunks in src and the result contains
+         * only chunks which satisfy whetherToKeep(chunk.chunkType)
+         * 
+         * @param src
+         * @return 
+         */
+        public List<PngChunk> filter(List<PngChunk> src) {
+            List<PngChunk> dst = new ArrayList<>();
+            
+            for(PngChunk chunk : src) {
+                if(whetherToKeep(chunk.chunkType))
+                    dst.add(chunk);
+            }
+            
+            return dst;
+        }
+        
+        public PngChunk filterOnlyOne(List<PngChunk> src) throws ImageReadException {
+            List<PngChunk> filtered = filter(src);
+            
+            if(filtered.size() > 1) {
+                throw new ImageReadException("More than one present in filtered: " +
+                        "type is " + filtered.get(0).chunkType);
+            } else if(filtered.size() == 0) {
+                return null;
+            }
+            
+            return filtered.get(0);
+        }
+        
+    }
+    
+    public static class ChunkListSelector extends ChunkSelector {
+
+        private final ChunkType[] list;
+        
+        public ChunkListSelector(boolean whiteList, ChunkType[] list) {
+            super(whiteList);
+            this.list = list;
+        }
+        
+        public ChunkType[] getList() {
+            return Arrays.copyOf(list, list.length);
+        }
+        
+        @Override
+        public boolean whetherToKeep(ChunkType chunkType) {
```

Comment:
Method name looks different. I think it would normally appear in an API as `keep()`, or `shouldKeep()`, etc.